src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Frameworks/SupportedTargetFrameworksProvider.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.Build.Framework.XamlTypes;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using TOutput = Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<System.Collections.Generic.ICollection<Microsoft.VisualStudio.ProjectSystem.Properties.IEnumValue>>;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Frameworks
+{
+    /// <summary>
+    ///     Responsible for producing valid values for the TargetFramework property from a design-time build.
+    /// </summary>
+    [ExportDynamicEnumValuesProvider("SupportedTargetFrameworksEnumProvider")]
+    [AppliesTo(ProjectCapability.DotNet)]
+    internal class SupportedTargetFrameworksProvider : ChainedProjectValueDataSourceBase<ICollection<IEnumValue>>, IDynamicEnumValuesProvider, IDynamicEnumValuesGenerator
+    {
+        private readonly IProjectSubscriptionService _subscriptionService;
+
+        public SupportedTargetFrameworksProvider(
+            ConfiguredProject project,
+            IProjectSubscriptionService subscriptionService)
+            : base(project, synchronousDisposal: true, registerDataSource: false)
+        {
+            _subscriptionService = subscriptionService;
+        }
+
+        protected override IDisposable? LinkExternalInput(ITargetBlock<TOutput> targetBlock)
```

Comment:
Who is the "external input" and who is the "target" here?

I can follow that we're taking data from the `ProjectBuildRuleSource`, extracting the `SupportedTargetFrameworkAlias` items, and transforming them into a collection of `IEnumValue` but I don't have a sense of where that operation fits into the bigger picture.

Reply:
If you haven't done much data flow, then this entire chain won't make a lot of sense. We are basically just transforming subscription into IEnumValue, which we end up returning in GetListedValuesAsync.